### PR TITLE
A true mixin approach for subclassing db classes

### PR DESCRIPTION
**Pull request by nhoening**
See #16.

We need a clearer way for the use cases where the user doesn't want to simply use the tables as proposed by timely_beliefs, but has their own table names etc.

### CODE REVIEW

File: .flake8 [hunk 1]
```diff
@@ -0,0 +1,5 @@
+[flake8]
```
Comment (Flix6x) [inline .flake8:1]: We also have a setup.cfg file than has some flake8 settings. I would recommend keeping all these config settings in that file, so possible conflicts (e.g. file length) can be identified within a single file.

File: .isort.cfg [hunk 2]
```diff
@@ -0,0 +1,9 @@
+[settings]
```
Comment (Flix6x) [inline .isort.cfg:1]: I believe this would become [isort] if moved to our setup.cfg file.

File: pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,2 @@
+[tool.black]
```
Comment (Flix6x) [inline pyproject.toml:1]: Likely can be moved to config.cfg

File: timely_beliefs/__init__.py [hunk 4]
```diff
@@ -1,12 +1,12 @@
 # flake8: noqa
 
-from timely_beliefs.sensors.classes import Sensor, DBSensor
-from timely_beliefs.sources.classes import BeliefSource, DBBeliefSource
+from timely_beliefs.sensors.classes import DBSensor, Sensor  # isort:skip
+from timely_beliefs.sources.classes import BeliefSource, DBBeliefSource  # isort:skip
```
Comment (Flix6x) [inline timely_beliefs/__init__.py:4]: Why are these skipped?

File: timely_beliefs/utils.py [hunk 5]
```diff
@@ -21,10 +21,10 @@ def all_of_type(l: Sequence, element_type) -> bool:
 
 
 def replace_multi_index_level(
-    df: "classes.BeliefsDataFrame",  # noqa: F821
+    df: "classes.BeliefsDataFrame",
     level: str,
     index: pd.Index,
-    intersection: bool = False,
+    intersection: bool = False,  # noqa: F821
```
Comment (Flix6x) [inline timely_beliefs/utils.py:27]: Is this noqa still needed?

File: timely_beliefs/sources/classes.py [hunk 6]
```diff
@@ -30,39 +33,29 @@ def __lt__(self, other):
...
+    # overwriting name as db field
     name = Column(String(120), nullable=False, default="")
 
     def __init__(self, name: str):
         BeliefSource.__init__(self, name)
-        Base.__init__(self)
 
     def __repr__(self):
-        return "<DBBeliefSource %s (%s)>" % (self.id, self.name)
+        return "<BeliefSource %s (%s)>" % (self.id, self.name)
```
Comment (Flix6x) [inline timely_beliefs/sources/classes.py:49]: Why the change?

File: timely_beliefs/tests/test_db_subclassing.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    assert session.query(RatedSource).filter(RatedSource.rating == 5).count() == 1
+
+    # We made one with default rating (0) and in conftest two are made in advance
+    assert session.query(RatedSource).filter(RatedSource.rating == 0).count() == 3
+
+
+class RatedSourceInCustomTable(Base, BeliefSourceDBMixin):
+    """ A custom db class for representing belief sources.
+    BeliefSource properties are added via the Mixin, so the Base is added by us.
+    Here, we can change the table name.
```
Comment (Flix6x) [inline timely_beliefs/tests/test_db_subclassing.py:45]: Great. Please also add some of this information to the main user documentation.

File: timely_beliefs/beliefs/classes.py [hunk 8]
```diff
@@ -127,58 +126,45 @@ def source_id(self):
...
-        if self.__name__ == "DBTimedBelief":
-            return {
-                "polymorphic_on": self.type,
-                "polymorphic_identity": "DBTimedBelief",
-            }
-        else:
-            return {"polymorphic_identity": self.__name__}
+    def sensor_id(cls):
+        return Column(
+            Integer(), ForeignKey("sensor.id", ondelete="CASCADE"), primary_key=True
```
Comment (Flix6x) [inline timely_beliefs/beliefs/classes.py:157]: So when the sensor is deleted, the beliefs are deleted, too? If so, please add that to the docs.

File: timely_beliefs/beliefs/classes.py [hunk 9]
```diff
@@ -127,58 +126,45 @@ def source_id(self):
...
+        )
 
-    def __init__(
-        self, sensor: DBSensor, source: DBBeliefSource, value: float, **kwargs
-    ):
-        TimedBelief.__init__(self, sensor, source, value, **kwargs)
+    @declared_attr
+    def source_id(cls):
+        return Column(Integer, ForeignKey("belief_source.id"), primary_key=True)
```
Comment (Flix6x) [inline timely_beliefs/beliefs/classes.py:162]: So when the source is deleted, the beliefs are not deleted? But the DBTimedeBelief class seems to suggest orphans will be deleted for both sensors AND sources.

File: timely_beliefs/__init__.py (hunk 4 above)
Comment (nhoening) [reply timely_beliefs/__init__.py:4]: The import order is very important here. The utils need these high-level classes, but isort would put them on top, due some sort order (I believe alphabetical)

File: timely_beliefs/utils.py (hunk 5 above)
Comment (nhoening) [reply timely_beliefs/utils.py:27]: I'll try without

File: timely_beliefs/__init__.py (hunk 4 above)
Comment (nhoening) [reply timely_beliefs/__init__.py:4]: isort has its sorting assumption (I believe alphabetical). In this case, itś crucial that the main classes are there first, as the utils rely on them.

File: timely_beliefs/sources/classes.py (hunk 6 above)
Comment (nhoening) [reply timely_beliefs/sources/classes.py:49]: Good catch

File: timely_beliefs/tests/test_db_subclassing.py (hunk 7 above)
Comment (nhoening) [reply timely_beliefs/tests/test_db_subclassing.py:45]: I've done so. Please take a look.

File: .flake8 (hunk 1 above)
Comment (nhoening) [reply .flake8]: That's true, I'll delete .flake8. For the others, I'm afraid that there is not one way to do everything in one file. 

flake8 can live in pyproject.toml and blakc/isort can go to setup.cfg I believe. I did improve the situation a bit. pyproject.tol is the preferred way forward it seems.
